File: activities/activities/leases.py
"""Session-directory leases — coordinates concurrent tool-call access to the
shared session filesystem PV (docs/components/session-filesystem.md,
"Resolved: Lease Renewal, Not Fixed TTL"; docs/components/state-layer.md's
`session_filesystem_leases` table).

Granularity is the session/subagent-*directory* level (not per-file — the
docs' own deliberate starting point, refined later only if real contention
patterns warrant it), so `path` here is exactly a `ids.session_fs_path(...)`
value. A lease is renewed periodically by whoever holds it (paired with the
same heartbeat tick a tool activity already does for cancellation — see
tools.py's `ToolContext.tick()`), not granted for a fixed TTL up front: it
only expires if renewal genuinely stops, mirroring the project's existing
"no ABANDON" cooperative-cancellation philosophy rather than introducing a
second, inconsistent timeout model.

`content_hash`/`last_writer` on the table are intentionally left unused by
this module — they belong to the future subagent merge-back mechanism
(session-filesystem.md's conflict-detection design), not to basic
acquire/renew/release.
"""

from __future__ import annotations

import asyncpg


async def acquire_or_renew(
    pool: asyncpg.Pool,
    session_key: str,
    path: str,
    holder_id: str,
    ttl_seconds: float,
) -> bool:
    """Acquire `path` for `holder_id`, or renew it if already held by the same
    holder. Fails (returns False, no error) if genuinely held by a different,
    non-expired holder — the caller decides whether/how to wait and retry;
    this function never blocks.
    """
    row = await pool.fetchrow(
        """
        INSERT INTO session_filesystem_leases (session_key, path, holder_id, expires_at)
        VALUES ($1, $2, $3, now() + make_interval(secs => $4))
        ON CONFLICT (session_key, path) DO UPDATE
            SET holder_id = EXCLUDED.holder_id,
                acquired_at = CASE WHEN session_filesystem_leases.holder_id = EXCLUDED.holder_id
                                   THEN session_filesystem_leases.acquired_at ELSE now() END,
                expires_at = EXCLUDED.expires_at
            WHERE session_filesystem_leases.holder_id = EXCLUDED.holder_id
               OR session_filesystem_leases.expires_at < now()
        RETURNING holder_id
        """,
        session_key,
        path,
        holder_id,
        ttl_seconds,
    )
    return row is not None and row["holder_id"] == holder_id


async def release(pool: asyncpg.Pool, session_key: str, path: str, holder_id: str) -> None:
    """Release a lease this holder currently holds. A no-op if it was already
    reclaimed by someone else after expiring — never errors on that."""
    await pool.execute(
        "DELETE FROM session_filesystem_leases WHERE session_key = $1 AND path = $2 AND holder_id = $3",
        session_key,
        path,
        holder_id,
    )

File: workflows/cmd/gateway/leases.go
package main

import (
	"context"
	"errors"
	"time"

	"github.com/jackc/pgx/v5"
)

// acquireOrRenewConnectionLease implements docs/components/gateway.md's
// "Resolved: Connection Leasing" — reuses session_filesystem_leases/
// leases.py's exact renewal-based compare-and-swap pattern verbatim, ported
// to Go. Keyed (platform, connectionID) rather than platform alone
// (2026-08-25 correction) — a tenant can run more than one connection of the
// same platform kind (e.g. two Discord bots), each needing its own lease row
// since each is a physically independent live socket. connectionID is
// always a platform-provided stable identity (Discord: the bot's own user
// id via GET /users/@me), resolved by the caller before this is ever called.
//
// Never blocks: returns (false, nil) — not an error — when the lease is
// genuinely held by someone else and hasn't expired. The caller decides
// whether/how to retry (docs/components/gateway.md: "it simply doesn't open
// a live connection this cycle and retries later — no error, no
// coordination protocol beyond the lease itself").
func (s *server) acquireOrRenewConnectionLease(ctx context.Context, platform, connectionID, holderID string, ttl time.Duration) (bool, error) {
	var gotHolder string
	err := s.pool.QueryRow(ctx, `
		INSERT INTO gateway_connection_leases (platform, connection_id, holder_id, expires_at)
		VALUES ($1, $2, $3, now() + make_interval(secs => $4))
		ON CONFLICT (platform, connection_id) DO UPDATE
			SET holder_id = EXCLUDED.holder_id,
			    acquired_at = CASE WHEN gateway_connection_leases.holder_id = EXCLUDED.holder_id
			                       THEN gateway_connection_leases.acquired_at ELSE now() END,
			    expires_at = EXCLUDED.expires_at
			WHERE gateway_connection_leases.holder_id = EXCLUDED.holder_id
			   OR gateway_connection_leases.expires_at < now()
		RETURNING holder_id
	`, platform, connectionID, holderID, ttl.Seconds()).Scan(&gotHolder)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			// The WHERE clause excluded every candidate row — genuinely
			// held by someone else, not expired. Not an error.
			return false, nil
		}
		return false, err
	}
	return gotHolder == holderID, nil
}

// releaseConnectionLease releases a lease this holder currently holds. A
// no-op if it was already reclaimed by someone else after expiring — never
// errors on that, mirroring leases.py's release exactly.
func (s *server) releaseConnectionLease(ctx context.Context, platform, connectionID, holderID string) error {
	_, err := s.pool.Exec(ctx,
		"DELETE FROM gateway_connection_leases WHERE platform = $1 AND connection_id = $2 AND holder_id = $3",
		platform, connectionID, holderID,
	)
	return err
}
